linked_list/linked_list.py

```diff
@@ -1,75 +1,181 @@
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(self, value):
-        pass
+
+        if self.head == None:
+            self.head = Node(value)
+        else:
+            new_node = Node(value)
+            new_node.next = self.head
+            self.head = new_node
+            return self.head
+
+        # if self.head == None:
+        #     return None 
+
+        # self.head = Node(value, self.head)
+        # self.length += 1
+        # return self.head
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        if self.head == None:
+            return False
+
+        if self.head.value == value:
+            return True
+
+        current = self.head
+        while current:
+            if current.value == value:
+                return True
+            current = current.next
+        
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+
+        length_counter = 0
+        current = self.head
+        while (current != None):
+            length_counter += 1
+            current = current.next
+
+        return length_counter
+
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+
+        # if self.head == None:
+        #     return None 
+
+        if self.length() <= index:
+            return None
+
+        current_value = 0 
+        current_node = self.head
+        while current_node:
+            if current_value == index:
+                return current_node.value
+            current_node = current_node.next
+            current_value += 1
+
+        return current_node.value
+        
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        if self.head == None:
+            return None
+
+        current = self.head
+        while current.next:
+            current = current.next
+        return current.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)

```

Comment:
⏱ Time complexity is going to be O(n) since you start at the head node and have to traverse the entire list.